src/lm/lm_chainer.py

```diff
@@ -115,32 +112,124 @@ def predict(self, state, x):
             state, z = self.predictor(state, x)
             return state, F.log_softmax(z).data
 
+    def final(self, state):
+        """Predict final log probabilities for given state using the predictor
+
+        Returns:
+            cupy/numpy array: log probability vector
+
+        """
+        if hasattr(self.predictor, 'final'):
+            return self.predictor.final(state)
+        else:
+            return 0.
+
 
 # Definition of a recurrent net for language modeling
 class RNNLM(chainer.Chain):
 
-    def __init__(self, n_vocab, n_units):
+    def __init__(self, n_vocab, n_layers, n_units):
         super(RNNLM, self).__init__()
         with self.init_scope():
-            self.embed = L.EmbedID(n_vocab, n_units)
-            self.l1 = L.StatelessLSTM(n_units, n_units)
-            self.l2 = L.StatelessLSTM(n_units, n_units)
+            self.embed = DL.EmbedID(n_vocab, n_units)
+            self.lstm = chainer.ChainList(
+                *[L.StatelessLSTM(n_units, n_units) for _ in range(n_layers)])
             self.lo = L.Linear(n_units, n_vocab)
 
         for param in self.params():
             param.data[...] = np.random.uniform(-0.1, 0.1, param.data.shape)
+        self.n_layers = n_layers
 
     def __call__(self, state, x):
         if state is None:
-            state = {'c1': None, 'h1': None, 'c2': None, 'h2': None}
-        h0 = self.embed(x)
-        c1, h1 = self.l1(state['c1'], state['h1'], F.dropout(h0))
-        c2, h2 = self.l2(state['c2'], state['h2'], F.dropout(h1))
-        y = self.lo(F.dropout(h2))
-        state = {'c1': c1, 'h1': h1, 'c2': c2, 'h2': h2}
+            state = {'c': [None] * self.n_layers, 'h': [None] * self.n_layers}
+        h = [None] * self.n_layers
+        c = [None] * self.n_layers
+        emb = self.embed(x)
+        c[0], h[0] = self.lstm[0](state['c'][0], state['h'][0], F.dropout(emb))
+        for n in six.moves.range(1, self.n_layers):
+            c[n], h[n] = self.lstm[n](state['c'][n], state['h'][n], F.dropout(h[n - 1]))
+        y = self.lo(F.dropout(h[-1]))
+        state = {'c': c, 'h': h}
         return state, y
 
 
+class BPTTUpdater(training.updaters.StandardUpdater):
```

Comment:
This class seems to be quite similar to the StandardUpdater. I feel that we may just use the StandardUpdater, and we don't have to use this kind of custom updater. But maybe I'm wrong.

Reply:
StandardUpdater does not have a loop for tokens in a given sequence, so we need to use this custom updater like the chainer's ptb example. Maybe it's possible to use the standard updater by changing `ClassifierWithState` class to include the loop. I think I can work on that, but need a bit more time. So it's better to do it in a different PR.